File: 백준/Platinum/7578. 공장/공장.py
import sys
input = sys.stdin.readline

n = int(input())
dic = {}
for i in input().split():
    dic[i] = len(dic)

seg = [0]*(n*2+1)

def modify(list,num,v):
    index = n + num
    while index>0:
        list[index]+=v
        index>>=1

def search(list,l,r):
    if l==r:
        return list[N + 1 + l]
    l = n + l
    r = n + r
    res = 0
    while l<=r:
        if l==r:
            res+=list[l]
            break
        if l%2==1:
            res+=list[l]
            l+=1
        if r%2==0:
            res+=list[r]
            r-=1
        l>>=1
        r>>=1
    return res
result = 0
for i in input().split():
    num = dic[i]
    result += search(seg,num,n)
    modify(seg,num,1)
print(result)

File: 백준/Platinum/7578. 공장/공장.java
import java.util.*;

public class Main {
    static int[] seg;
    static int n;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        n = sc.nextInt();
        sc.nextLine(); // 개행 제거

        // 첫 번째 문자열 입력 (dic 만들기용)
        String[] firstInput = sc.nextLine().split(" ");
        Map<String, Integer> dic = new HashMap<>();
        for (String s : firstInput) {
            dic.putIfAbsent(s, dic.size());
        }

        // 세그먼트 트리 초기화
        seg = new int[n * 2 + 1];

        // 두 번째 문자열 입력
        String[] secondInput = sc.nextLine().split(" ");
        long result = 0;
        for (String s : secondInput) {
            int num = dic.get(s);
            result += search(num, n - 1);
            modify(num, 1);
        }

        System.out.println(result);
    }

    // 세그먼트 트리 갱신 함수
    static void modify(int idx, int val) {
        int index = n + idx;
        while (index > 0) {
            seg[index] += val;
            index >>= 1;
        }
    }

    // 세그먼트 트리 구간합 조회 함수
    static int search(int l, int r) {
        int res = 0;
        l += n;
        r += n;
        while (l <= r) {
            if (l == r) {
                res += seg[l];
                break;
            }
            if ((l & 1) == 1) {
                res += seg[l++];
            }
            if ((r & 1) == 0) {
                res += seg[r--];
            }
            l >>= 1;
            r >>= 1;
        }
        return res;
    }
}
